File: cpp/src/IceSSL/uwp/AcceptorI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <IceSSL/uwp/AcceptorI.h>
#include <IceSSL/uwp/TransceiverI.h>
#include <IceSSL/uwp/EndpointI.h>

#include <Ice/ProtocolInstance.h>
#include <Ice/LocalException.h>
#include <Ice/LoggerUtil.h>
#include <Ice/Exception.h>
#include <Ice/Properties.h>
#include <IceUtil/StringUtil.h>

using namespace std;
using namespace Ice;
using namespace IceSSL;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Storage::Streams;
using namespace Windows::Networking;
using namespace Windows::Networking::Sockets;

IceUtil::Shared* IceSSL::upCast(AcceptorI* p) { return p; }

IceInternal::NativeInfoPtr
IceSSL::AcceptorI::getNativeInfo()
{
    return this;
}

void
IceSSL::AcceptorI::setCompletedHandler(IceInternal::SocketOperationCompletedHandler^ handler)
{
    _completedHandler = handler;
}

void
IceSSL::AcceptorI::close()
{
    IceUtil::Mutex::Lock lock(_mutex);
    if(_acceptPending)
    {
        assert(_accepted.empty());
        _completedHandler(IceInternal::SocketOperationRead);
        _acceptPending = false;
    }
    else if(!_accepted.empty())
    {
        for(deque<StreamSocket^>::const_iterator p = _accepted.begin(); p != _accepted.end(); ++p)
        {
            IceInternal::closeSocket(*p);
        }
        _accepted.clear();
    }

    if(_fd != INVALID_SOCKET)
    {
        IceInternal::closeSocketNoThrow(_fd);
        _fd = INVALID_SOCKET;
    }
}

IceInternal::EndpointIPtr
IceSSL::AcceptorI::listen()
{
    try
    {
        const_cast<IceInternal::Address&>(_addr) = IceInternal::doBind(_fd, _addr);
    }
    catch(...)
    {
        _fd = INVALID_SOCKET;
        throw;
    }
    _endpoint = _endpoint->endpoint(this);
    return _endpoint;
}

void
IceSSL::AcceptorI::startAccept()
{
    assert(_fd != INVALID_SOCKET);

    //
    // If there are already sockets waiting to be accepted, we just
    // notify the selector that the acceptor is ready for acceting the
    // new socket. Otherwise, we set the _acceptPending flag, when a
    // new socket connection event is received, the message handler
    // will notify the selector.
    //
    IceUtil::Mutex::Lock lock(_mutex);
    assert(!_acceptPending);
    if(!_accepted.empty())
    {
        _completedHandler(IceInternal::SocketOperationRead);
    }
    else
    {
        _acceptPending = true;
    }
}

void
IceSSL::AcceptorI::finishAccept()
{
    //
    // Nothing to do, we just check there's at least one accepted
    // socket or the acceptor was closed.
    //
    IceUtil::Mutex::Lock lock(_mutex);
    assert(!_acceptPending && (!_accepted.empty() || _fd == INVALID_SOCKET));
}

IceInternal::TransceiverPtr
IceSSL::AcceptorI::accept()
{
    if(_fd == INVALID_SOCKET) // Acceptor closed.
    {
        assert(_accepted.empty());
        throw SocketException(__FILE__, __LINE__);
    }

    StreamSocket^ fd;
    {
        IceUtil::Mutex::Lock lock(_mutex);
        assert(!_accepted.empty());
        fd = _accepted.front();
        _accepted.pop_front();
    }

    return new TransceiverI(_instance, fd, true);
}

string
IceSSL::AcceptorI::protocol() const
{
    return _instance->protocol();
}

string
IceSSL::AcceptorI::toString() const
{
    return IceInternal::addrToString(_addr);
}

string
IceSSL::AcceptorI::toDetailedString() const
{
    ostringstream os;
    os << "local address = " << toString();
    vector<string> intfs = IceInternal::getHostsForEndpointExpand(IceInternal::inetAddrToString(_addr), _instance->protocolSupport(), true);
    if(!intfs.empty())
    {
        os << "\nlocal interfaces = ";
        os << IceUtilInternal::joinString(intfs, ", ");
    }
    return os.str();
}

int
IceSSL::AcceptorI::effectivePort() const
{
    return IceInternal::getPort(_addr);
}

IceSSL::AcceptorI::AcceptorI(const EndpointIPtr& endpoint,
                             const IceInternal::ProtocolInstancePtr& instance,
                             const string& host,
                             int port) :
    _endpoint(endpoint),
    _instance(instance),
    _addr(getAddressForServer(host, port, _instance->protocolSupport(), instance->preferIPv6())),
    _acceptPending(false)
{
    _fd = ref new StreamSocketListener();

    safe_cast<StreamSocketListener^>(_fd)->ConnectionReceived +=
        ref new TypedEventHandler<StreamSocketListener^, StreamSocketListenerConnectionReceivedEventArgs^>(
            [=](StreamSocketListener^, StreamSocketListenerConnectionReceivedEventArgs^ args)
                {
                    queueAcceptedSocket(args->Socket);
                });
}

IceSSL::AcceptorI::~AcceptorI()
{
    assert(_fd == INVALID_SOCKET);
}

void
IceSSL::AcceptorI::queueAcceptedSocket(StreamSocket^ socket)
{
    IceUtil::Mutex::Lock lock(_mutex);
    if(_fd == INVALID_SOCKET) // Acceptor was closed.
    {
        IceInternal::closeSocket(socket);
        return;
    }
    _accepted.push_back(socket);

    //
    // If the acceptor is waiting for a socket to be accepted, notify
    // the selector that the acceptor is ready for "read". This will
    // in turn caused finishAccept() and accept() to be called by the
    // thread pool. If the acceptor isn't ready to accept the socket,
    // it is just queued, when startAccept is called it will be dequed.
    //
    if(_acceptPending)
    {
        _completedHandler(IceInternal::SocketOperationRead);
        _acceptPending = false;
    }
}

File: java/src/IceBT/src/main/java/com/zeroc/IceBT/AcceptorI.java
// **********************************************************************
//
// Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package com.zeroc.IceBT;

import com.zeroc.IceInternal.Acceptor;
import com.zeroc.IceInternal.ReadyCallback;
import com.zeroc.IceInternal.SocketOperation;
import com.zeroc.IceInternal.Transceiver;
import com.zeroc.Ice.SocketException;

import android.bluetooth.BluetoothSocket;
import android.bluetooth.BluetoothServerSocket;
import java.util.UUID;

final class AcceptorI implements Acceptor
{
    @Override
    public java.nio.channels.ServerSocketChannel fd()
    {
        return null;
    }

    @Override
    public void setReadyCallback(ReadyCallback callback)
    {
        _readyCallback = callback;
    }

    @Override
    public void close()
    {
        synchronized(this)
        {
            _closed = true;
        }

        if(_socket != null)
        {
            try
            {
                _socket.close(); // Wakes up the thread blocked in accept().
            }
            catch(Exception ex)
            {
                // Ignore.
            }
        }
        if(_thread != null)
        {
            try
            {
                _thread.join();
            }
            catch(Exception ex)
            {
                // Ignore.
            }
        }
    }

    @Override
    public com.zeroc.IceInternal.EndpointI listen()
    {
        UUID uuid = null;
        try
        {
            uuid = UUID.fromString(_uuid);
        }
        catch(IllegalArgumentException ex)
        {
            throw new SocketException(ex);
        }

        try
        {
            //
            // We always listen using the "secure" method.
            //
            _socket = _instance.bluetoothAdapter().listenUsingRfcommWithServiceRecord(_name, uuid);
        }
        catch(java.io.IOException ex)
        {
            throw new SocketException(ex);
        }

        //
        // Use a helper thread to perform the blocking accept() calls.
        //
        _thread = new Thread()
        {
            public void run()
            {
                runAccept();
            }
        };
        _thread.start();

        return _endpoint;
    }

    @Override
    public synchronized Transceiver accept()
    {
        if(_exception != null)
        {
            throw new SocketException(_exception);
        }

        //
        // accept() should only be called when we have at least one socket ready.
        //
        assert(!_pending.isEmpty());

        BluetoothSocket socket = _pending.pop();

        //
        // Update our status with the thread pool.
        //
        _readyCallback.ready(SocketOperation.Read, !_pending.isEmpty());

        return new TransceiverI(_instance, socket, _uuid, _adapterName);
    }

    @Override
    public String protocol()
    {
        return _instance.protocol();
    }

    @Override
    public String toString()
    {
        StringBuffer s = new StringBuffer("local address = ");
        s.append(_instance.bluetoothAdapter().getAddress());
        return s.toString();
    }

    @Override
    public String toDetailedString()
    {
        StringBuffer s = new StringBuffer(toString());
        if(!_name.isEmpty())
        {
            s.append("\nservice name = '");
            s.append(_name);
            s.append("'");
        }
        if(!_uuid.isEmpty())
        {
            s.append("\nservice uuid = ");
            s.append(_uuid);
        }
        return s.toString();
    }

    AcceptorI(EndpointI endpoint, Instance instance, String adapterName, String uuid, String name)
    {
        _endpoint = endpoint;
        _instance = instance;
        _adapterName = adapterName;
        _name = name;
        _uuid = uuid;

        _pending = new java.util.Stack<BluetoothSocket>();
        _closed = false;
    }

    private void runAccept()
    {
        try
        {
            while(true)
            {
                BluetoothSocket socket = _socket.accept();
                synchronized(this)
                {
                    _pending.push(socket);

                    //
                    // Notify the thread pool that we are ready to "read". The thread pool will invoke accept()
                    // and we can return a new transceiver.
                    //
                    _readyCallback.ready(SocketOperation.Read, true);
                }
            }
        }
        catch(Exception ex)
        {
            synchronized(this)
            {
                if(!_closed)
                {
                    _exception = ex;
                    _readyCallback.ready(SocketOperation.Read, true);
                }
            }
        }

        //
        // Close any remaining incoming sockets that haven't been accepted yet.
        //
        java.util.Stack<BluetoothSocket> pending;
        synchronized(this)
        {
            pending = _pending;
            _pending = null;
        }

        for(BluetoothSocket s : pending)
        {
            try
            {
                s.close();
            }
            catch(Exception ex)
            {
                // Ignore.
            }
        }
    }

    private EndpointI _endpoint;
    private Instance _instance;
    private String _adapterName;
    private String _name;
    private String _uuid;
    private ReadyCallback _readyCallback;
    private BluetoothServerSocket _socket;
    private java.util.Stack<BluetoothSocket> _pending;
    private Thread _thread;
    private Exception _exception;
    private boolean _closed;
}
